Add waypoint handling

bsudekum opened the pull request:
To do:

- [x] Add waypoint handling to example app
- [x] Improve instructions upon approaching leg destination
- ~~[x] Add UI elements for arriving and leaving waypoint~~

/cc @1ec5 @frederoni @cammace @ericrwolfe

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -72,14 +72,14 @@ extension RouteTableViewController: UITableViewDelegate, UITableViewDataSource {
     
     func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return routeController.routeProgress.currentLeg.steps.count
+        return routeController.routeProgress.allSteps.count
```
Instead of flattening the list of steps, we should have multiple sections (instead of just one), each section listing the steps for a single leg.

bsudekum commented:
Some notes, 

* I deprecated `destination`. It was never really needed and with waypoints, things get a little more confusing if there is a destination option. 
* `didArriveAt` changed to now return the route step since there really is not a destination annotation anymore. It also will only return when the user arrives at the end of the route. However, we could change this to return at ever waypoint.
* In terms of how this will fit into existing code bases, I think it would just work. There are improvements we could make to how waypoints are handled (adding a UI namely), but as a first step, I think this is good and would unblock any users looking for waypoint support.

/cc @1ec5 @frederoni

1ec5 commented on MapboxCoreNavigation/RouteController.swift:
```diff
@@ -257,7 +258,17 @@ extension RouteController: CLLocationManagerDelegate {
+                routeProgress.remainingWaypoints.count > 1 {
+                routeProgress.legIndex += 1
+                routeProgress.currentLegProgress.stepIndex = 0
+                routeProgress.currentLegProgress.alertUserLevel = .depart
```
Changing `legIndex` automatically [resets](https://github.com/mapbox/mapbox-navigation-ios/blob/5d78f52e52d2e1e847336890a44bfaae6505f653/MapboxCoreNavigation/RouteProgress.swift#L135) `currentLegProgress`.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -124,8 +124,10 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
      on the destination of your route. The last coordinate of the route will be
      used if no destination is given.
     */
+    @available(*, deprecated, message: "Destination is no longer support nor necessary. A destination annotation will automatically be added to map given the route.")
```
Note that this shouldn’t have much of an impact on most developers: it isn’t possible to customize this annotation (#213) other than changing its title, subtitle, and coordinates, and none of those properties are likely to be modified by the developer anyways.

1ec5 replied:
s/support/supported/

bsudekum replied:
@1ec5 it was actually added in https://github.com/mapbox/mapbox-navigation-ios/pull/268

1ec5 replied:
Ah, I missed that we’re exposing `navigationMapView(_:imageFor:)`. In any case, that method will continue to work; it just won’t be possible to say something like `if annotation == navigationViewController.destination`.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -368,10 +369,11 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
-            annotation.coordinate = route.coordinates!.last!
-            destination = annotation
+        for leg in route.legs {
+            if let last = leg.steps.last {
```
This code only places an annotation at the last destination. Should we annotate the intermediate waypoints as well?

bsudekum replied:
@1ec5 hrm, no this annotates the last step of each leg. Or at least it's supposed to.

1ec5 replied:
You’re right, I misread this line.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -368,10 +369,11 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
-            let annotation = MGLPointAnnotation()
-            annotation.coordinate = route.coordinates!.last!
-            destination = annotation
+        for leg in route.legs {
```
If we’re only looking at the last step, there’s no need to iterate over all the steps:

```swift
if let lastStep = route.legs.last?.steps.last {
```

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -368,10 +369,11 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
+        for leg in route.legs {
+            if let last = leg.steps.last {
+                let annotation = MGLPointAnnotation()
+                annotation.coordinate = last.coordinates!.last!
```
The step’s `destination` is a Waypoint whose `coordinates` property corresponds to the location we’d want to annotate.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -20,7 +20,7 @@ public protocol NavigationViewControllerDelegate {
      Called when the user arrives at the destination.
      */
-    @objc optional func navigationViewController(_ navigationViewController : NavigationViewController, didArriveAt destination: MGLAnnotation)
+    @objc optional func navigationViewController(_ navigationViewController : NavigationViewController, didArriveAt destination: RouteStep)
```
RouteStep isn’t a destination. Arriving at a step means something different than arriving at a destination. I agree that MGLAnnotation is a poor fit for this method, but then we should pass in a Waypoint, not a RouteStep.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -319,9 +319,10 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
-            navigationDelegate?.navigationViewController?(self, didArriveAt: destination)
+
+        if routeProgress.currentLegProgress.alertUserLevel == .arrive,
+            routeProgress.remainingWaypoints.count == 0 {
```
Should the developer be notified when the user reaches an intermediate waypoint? I could see use cases for only responding to the final arrival and for responding to intermediate arrivals as well. Perhaps the delegate method should have an additional argument, `numberOfWaypointsRemaining`, to allow the developer to distinguish between the two cases.

bsudekum replied:
@1ec5 I was thinking about this but wanted to keep the API the same. I think if we want to notify them when they reach a waypoint, we should either rename the method or add a new method for waypoints.

1ec5 replied:
The method’s current name implies that it’s called whenever the user reaches a waypoint once we added support for multiple waypoints. Why else would it take a second argument?

How about we add a new method, `navigationViewControllerDidArrive(_:)`, that gets called after `navigationViewController(_:didArriveAt:)` when arriving at the final waypoint?

bsudekum replied:
Changed it to a single method that returns the waypoint the user is arriving at for each and every waypoint.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,25 @@ class RouteTableViewController: UIViewController {
+    }
+    
+    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+        return "Leg \(section + 1)"
```
The leg number is unimportant to the user. Instead, we should title the section with either of the following formats, preferably the first one:

* “Biloxi to Mobile” (see `leg.source`, `leg.destination`)
* “I-10” (see `leg.name`)

1ec5 replied:
If there is only one leg, there’s no need for a section header.

1ec5 replied:
Intermediate waypoints returned by the Directions API generally lack names. In that case, the second format is probably the best we can do, unless we’re OK with adding a dependency on the Geocoding API.

1ec5 replied:
Blocked by mapbox/MapboxDirections.swift#135.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,25 @@ class RouteTableViewController: UIViewController {
+        
         
-        cell.step = leg.steps[indexPath.row]
+         cell.step = legs[indexPath.section].steps[indexPath.row]
```
Nit: funky indentation.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,25 @@ class RouteTableViewController: UIViewController {
+         cell.step = legs[indexPath.section].steps[indexPath.row]
         
-        if routeController.routeProgress.currentLegProgress.stepIndex + 1 > indexPath.row {
+        if  routeController.routeProgress.legIndex + 1 > indexPath.section && routeController.routeProgress.currentLegProgress.stepIndex + 1 > indexPath.row  {
```
If the user is on step 5 of leg 1, only the first five steps of section 0 are dimmed out.

bsudekum replied:
@1ec5 I think that's how it should work though. Are you saying the current step should be dimmed?

1ec5 replied:
If leg 0 has 10 steps, leg 1 has 10 steps, and the user is on step 5 of leg 2 which has 10 steps, why would all three steps be halfway dimmed?

1ec5 replied:
Actually, this code is incorrect but in a different way than I first thought. This code is equivalent to: `if currentLegIndex >= section && currentLegProgress.stepIndex >= row { dim the cell }`. Now if the user is on step 5 of leg 2, the only dimmed cells are those that meet _both_ the following criteria:

1. the section number is at least 2; and
2. the row number is at least 5

In other words:

* Sections 0 and 1 are completely lit, because they fail criteria (1). 🙅‍♂️
* Section 2 is dimmed up to row 4 and lit beginning with row 5. 👍
* Sections 3, 4, etc. are each dimmed up to row 4 but lit beginning with row 5, because rows 5 and beyond fail criteria (2). 🙅‍♂️

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -72,7 +72,17 @@ extension RouteTableViewController: UITableViewDelegate, UITableViewDataSource {
+            return nil
+        }
+        
+        let legToFrom = routeController.routeProgress.route.legs[section].description.components(separatedBy: ", ")
```
`RouteLeg.description` is equivalent to `RouteLeg.name`. `name` lists the main streets along the leg, not the leg’s origin and destination. For example, `name` might be “Market Street, I-10, Main Street”. If you want “Biloxi” and “Mobile”, you need to look at the `name` properties of RouteLeg’s `source` and `destination` properties, which won’t be very useful until mapbox/MapboxDirections.swift#135 is fixed.

bsudekum commented:
This is ready for another round of reviews

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -20,7 +20,7 @@ public protocol NavigationViewControllerDelegate {
      Called when the user arrives at the destination.
      */
-    @objc optional func navigationViewController(_ navigationViewController : NavigationViewController, didArriveAt destination: MGLAnnotation)
+    @objc optional func navigationViewController(_ navigationViewController : NavigationViewController, didArriveAt destination: Waypoint)
```
Should we also add `navigationViewControllerDidArrive(_:)` (with only one argument of type NavigationViewController)?

bsudekum replied:
I think one method is good enough here.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -332,6 +328,10 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
         tableViewController?.notifyAlertLevelDidChange()
         
+        if routeProgress.currentLegProgress.alertUserLevel == .arrive {
+            navigationDelegate?.navigationViewController?(self, didArriveAt: routeProgress.route.routeOptions.waypoints[routeProgress.legIndex + 1])
```
We should pass in the current route leg’s destination waypoint. Otherwise, if we pass in one of the waypoints used to request the route, its location won’t necessarily correspond to what the Directions API thinks the destination should be (after snapping or whatever).

1ec5 commented on MapboxNavigation/RouteMapViewController.swift:
```diff
@@ -76,7 +75,14 @@ class RouteMapViewController: UIViewController, PulleyPrimaryContentControllerDe
+        for leg in route.legs {
+            if let last = leg.steps.last {
+                let annotation = MGLPointAnnotation()
+                annotation.coordinate = last.coordinates!.last!
```
https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r121788876

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,37 @@ class RouteTableViewController: UIViewController {
+        }
+        
+        let sectionName = routeController.routeProgress.route.legs[section].name
+        let legToFrom = sectionName.components(separatedBy: ", ")
```
The suggestion in https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r121870552 was to use `source` and `destination`, not `name`.

bsudekum replied:
I added using source and destination but I also left this code in place for cases where we do not have source and destination.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,37 @@ class RouteTableViewController: UIViewController {
+        cell.step = legs[indexPath.section].steps[indexPath.row]
         
-        if routeController.routeProgress.currentLegProgress.stepIndex + 1 > indexPath.row {
+        if routeController.routeProgress.legIndex + 1 > indexPath.section && routeController.routeProgress.currentLegProgress.stepIndex + 1 > indexPath.row  {
```
https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r121869861

bsudekum replied:
@1ec5 having a hard time getting something to work here. Have any ideas how to accomplish this?

1ec5 replied:
We need to dim a cell for either of two reasons:

1. Either it represents a step that’s part of a leg that the user has already completed,
1. Or it represents a completed step in the current leg.

So the correct logic would be:

```swift
if indexPath.section < routeController.routeProgress.legIndex
    || (indexPath.section == routeProgress.legIndex
        && indexPath.row <= routeProgress.currentLegProgress.stepIndex) {
    // dim the cell
}
```

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -76,14 +76,20 @@ extension RouteTableViewController: UITableViewDelegate, UITableViewDataSource {
+        
+        let legToFrom = leg.name.components(separatedBy: ", ")
         if legToFrom.count == 2 {
             return "\(legToFrom[0]) to \(legToFrom[1])"
```
Since these are two road names, “Main Street **and** Market Street” would be clearer than “Main Street **to** Market Street”.

1ec5 commented:
Blocked by Project-OSRM/osrm-text-instructions.swift#24.

1ec5 commented on Examples/Swift/ViewController.swift:
```diff
@@ -51,61 +55,40 @@ class ViewController: UIViewController, MGLMapViewDelegate {
-    @IBAction func didTapClearmap(_ sender: Any) {
-        mapView.removeAnnotations(mapView.annotations ?? [])
+    @IBAction func replay(_ sender: Any) {
+        let bundle = Bundle(for: ViewController.self)
```
This class is inside the application target, so you can say simply `Bundle.main`. It’s only necessary to use `Bundle(for:)` inside a framework, since the framework isn’t the main bundle.

1ec5 commented on Examples/Swift/ViewController.swift:
```diff
@@ -239,39 +218,38 @@ class ViewController: UIViewController, MGLMapViewDelegate {
+}
+
+extension ViewController: WaypointConfirmationViewControllerDelegate {
+    func confirmationControllerDidConfirm(controller confirmationController: WaypointConfirmationViewController) {
```
The first part of this selector, `confirmationControllerDidConfirm`, already contains “confirmation controller”, so `controller` can be `_`.

This code path would be a lot cleaner with an unwind segue instead of the delegate pattern…

1ec5 commented on MapboxCoreNavigation/RouteProgress.swift:
```diff
@@ -107,7 +107,13 @@ open class RouteProgress: NSObject {
+     Intializes a new `RouteProgress`.
+     
+     - parameter route: The route to follow.
+     - parameter legIndex: Zero-based index indicating the current leg the user is on.
```
`legIndex` is the index of the leg to start on.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,43 @@ class RouteTableViewController: UIViewController {
+        
+        let legToFrom = leg.name.components(separatedBy: ", ")
+        if legToFrom.count == 2 {
+            return "\(legToFrom[0]) to \(legToFrom[1])"
```
Not sure why GitHub thinks https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r123045968 is outdated, but it’s still current.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,43 @@ class RouteTableViewController: UIViewController {
+        let sourceName = leg.source.name
+        let destinationName = leg.destination.name
+        
+        if let sourceName = sourceName?.nonEmptyString, let destinationName = destinationName?.nonEmptyString {
```
If the source is the user’s unnamed current location and the destination is a named location, we should say “Destination via Street Name and Street Name”.

If the source is a named location and the destination is unnamed for some reason, we should say “Street Name and Street Name”, same as below.

1ec5 commented:
Project-OSRM/osrm-text-instructions.swift#24 has landed, by the way.

1ec5 commented:
This PR fixes #204.

frederoni commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,41 @@ class RouteTableViewController: UIViewController {
         
-        if routeController.routeProgress.currentLegProgress.stepIndex + 1 > indexPath.row {
+        if indexPath.section < routeController.routeProgress.legIndex || (indexPath.section == routeController.routeProgress.legIndex && indexPath.row <= routeController.routeProgress.currentLegProgress.stepIndex) {
             cell.contentView.alpha = 0.4
```
Upcoming steps may appear faded out if a passed step cell is being reused.
Consider adding an else statement or more succinctly `cell.contentView.alpha = didPassStep ? 0.4 : 1`

1ec5 commented on Examples/Swift/ViewController.swift:
```diff
@@ -254,37 +262,6 @@ extension ViewController: NavigationViewControllerDelegate {
 
 extension ViewController: WaypointConfirmationViewControllerDelegate {
     func confirmationControllerDidConfirm(controller confirmationController: WaypointConfirmationViewController) {
```
Per https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r126295971, rename `controller` to `_`.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,41 @@ class RouteTableViewController: UIViewController {
+        if let destinationName = destinationName?.nonEmptyString, legToFrom.count > 1 {
+            return "\(destinationName) via \(legToFrom.joined(separator: " and "))"
+        } else if let sourceName = sourceName?.nonEmptyString, let destinationName = destinationName?.nonEmptyString {
+            return "\(sourceName) anz \(destinationName)"
```
Typo: and.

1ec5 replied:
We should use a localizable format string instead of string interpolation.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -68,20 +68,41 @@ class RouteTableViewController: UIViewController {
+        let legToFrom = leg.name.components(separatedBy: ", ")
+        
+        if let destinationName = destinationName?.nonEmptyString, legToFrom.count > 1 {
+            return "\(destinationName) via \(legToFrom.joined(separator: " and "))"
```
It would be easier to check for `leg.name` containing `, ` and replace `, ` with ` and `. But in any case, we should use a localizable format string instead of string interpolation.

1ec5 commented on MapboxNavigation/Resources/Base.lproj/Localizable.strings:
```diff
@@ -22,3 +22,8 @@
+"WAYPOINT_DESTINATION_VIA_WAYPOINTS_FORMAT" = "%1$@, via %2$@";
+
+/* Format for displaying start and endpoint for leg; 1 = source; 2 = destination */
+"WAYPOINT_SOURCE_DESTINATION_FORMAT" = "%1$@, and %2$@";
```
No comma needed, since there’s only two items in the list.

1ec5 commented:
It might be worth fixing Project-OSRM/osrm-text-instructions#61 as part of this effort, since the past perfect (“you have arrived”) stands out even more for intermediate waypoints. It would be a trivial change as far as I can tell.

frederoni commented on MapboxNavigation/RouteMapViewController.swift:
```diff
@@ -448,10 +440,52 @@ extension RouteMapViewController: MGLMapViewDelegate {
     }
     
     func showRouteIfNeeded() {
-        guard isViewLoaded && view.window != nil else { return }
```
UIAppearance styling of the route line should not work w/o this line and L:113

bsudekum commented:
Updated this:
* Added custom waypoint markers going from A -> Z
* Added `navigationViewController(navigationViewController:shouldIncrementLegWhenArrivingAtWaypoints:)`. By default, the leg index will be incremented automatically when the user arrives at a waypoint. If the developer wants to show a custom view when arriving at a waypoint, they can set this to false, show their UI and on their own increment leg index.

bsudekum commented on MapboxNavigation/RouteMapViewController.swift:
```diff
@@ -447,10 +443,52 @@ extension RouteMapViewController: MGLMapViewDelegate {
+        
+        if route.legs.count > 1 {
+            var features = [MGLPointFeature]()
+            let letters = (97...122).map({Character(UnicodeScalar($0))}).map { String(describing:$0).uppercased() }
```
Unsure how this will translate.

/cc @1ec5

frederoni commented on Examples/Swift/ViewController.swift:
```diff
@@ -121,7 +133,7 @@ class ViewController: UIViewController, MGLMapViewDelegate {
         
         _ = Directions.shared.calculate(options) { [weak self] (waypoints, routes, error) in
             guard error == nil else {
```
While you're at it, let's replace this with a `guard let` so we don't have to force unwrap the error.

frederoni commented on Examples/Swift/WaypointConfirmationViewController.swift:
```diff
@@ -1,14 +1,14 @@
 
 protocol WaypointConfirmationViewControllerDelegate: NSObjectProtocol {
-    func confirmationControllerDidConfirm(controller: WaypointConfirmationViewController)
+    func confirmationControllerDidConfirm(_: WaypointConfirmationViewController)
```
Let's keep the name but avoid the prefix so that the scope variable comes with a predefined name. (`_ controller`)

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -125,6 +129,56 @@ open class NavigationMapView: MGLMapView {
+        for (waypointIndex, waypoint) in routeProgress.remainingWaypoints.enumerated().dropLast() {
+            let feature = MGLPointFeature()
+            feature.coordinate = waypoint.coordinate
+            feature.attributes = [ "name": letters[waypointIndex] ]
```
```swift
feature.attributes = [
    "name": Character(UnicodeScalar(0x41 /* A */ + waypointIndex))
]
```

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -142,37 +142,17 @@ open class NavigationMapView: MGLMapView {
-            feature.attributes = [ "name": letters[waypointIndex] ]
-            features.append(feature)
-        }
+        let source = navigationMapDelegate?.navigationMapView?(self, shapeFor: routeProgress.remainingWaypoints) ?? shape(for: Array(routeProgress.remainingWaypoints.dropLast()))
```
Let’s give this variable and `s` below more descriptive names.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -75,12 +75,19 @@ public protocol NavigationViewControllerDelegate {
      If this method is unimplemented, the navigation map view draws the route line using an `MGLLineStyleLayer`.
      */
-    @objc optional func navigationMapView(_ mapView: NavigationMapView, routeStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
+    @objc optional func navigationMapView(_ mapView: NavigationMapView, routeWaypointCircleStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
```
The name of this method implies that the return value must be an MGLCircleStyleLayer, but that isn’t the case. A better name would be `navigationMapView(_:waypointStyleLayerWithIdentifier:source:)`.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -75,12 +75,19 @@ public protocol NavigationViewControllerDelegate {
+     
+     If this method is unimplemented, the navigation map view draws the route line using an `MGLLineStyleLayer`.
+     */
+    @objc optional func navigationMapView(_ mapView: NavigationMapView, routeWaypointSymbolStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
```
What if the developer needs more than just a single circle style layer and symbol style layer to represent the waypoint? It seems like we need a method that returns an array of style layers. NavigationMapView would need to keep track of the identifiers of the returned style layers so it can refer to them later on.

bsudekum replied:
Do you think we should follow the same logic with the route line since it is also made up of multiple parts?

1ec5 replied:
Possibly, but that would limit the developer’s options considerably. For example, what if the developer wants to mark the waypoint with a circle and two symbols – a hotel icon overlaid on the circle and a text label underneath?

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -75,12 +75,19 @@ public protocol NavigationViewControllerDelegate {
+     If this method is unimplemented, the navigation map view draws the route line using an `MGLLineStyleLayer`.
+     */
+    @objc optional func navigationMapView(_ mapView: NavigationMapView, routeWaypointSymbolStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
     
```
What if the waypoint needs to be represented by an annotation, to match the main map in their application? There should be alternative, optional methods that mirror `MGLMapViewDelegate.mapView(_:imageForAnnotation:)` and `MGLMapViewDelegate.mapView(_:viewForAnnotation:)`. When either `navigationMapView(_:imageForAnnotation:waypoint:)` or `navigationMapView(_:viewForAnnotation:waypoint:)` is implemented, NavigationMapView would take care of adding an MGLPointAnnotation instead of a source and style layer.

/ref #114

1ec5 replied:
#1993

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -34,6 +34,9 @@ public protocol NavigationViewControllerDelegate {
     @objc(navigationViewController:shouldRerouteFromLocation:)
     optional func navigationViewController(_ navigationViewController: NavigationViewController, shouldRerouteFrom location: CLLocation) -> Bool
     
+    @objc(navigationViewController:shouldIncrementLegWhenArrivingAtWaypoints:)
```
This public method needs a documentation comment. (Same goes for the corresponding method on RouteController.)

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -34,6 +34,9 @@ public protocol NavigationViewControllerDelegate {
     optional func navigationViewController(_ navigationViewController: NavigationViewController, shouldRerouteFrom location: CLLocation) -> Bool
     
+    @objc(navigationViewController:shouldIncrementLegWhenArrivingAtWaypoints:)
+    optional func navigationViewController(_ navigationViewController: NavigationViewController, shouldIncrementLegWhenArrivingAtWaypoints waypoint: Waypoint) -> Bool
```
I’m confused by the `shouldIncrementLegWhenArrivingAtWaypoints` argument: it makes the method sound like there’s only one setting for all waypoints in the route, yet this method is called once for each waypoint. Did you mean to make the argument name singular, i.e. `shouldIncrementLegWhenArrivingAtWaypoint`? If so, the Swift name should be `navigationViewController(_:shouldIncrementLegWhenArrivingAt:)` while the Objective-C name would be `-navigationViewController:shouldIncrementLegWhenArrivingAtWaypoint:`.

1ec5 commented on MapboxNavigation/RouteMapViewController.swift:
```diff
@@ -324,8 +322,12 @@ extension RouteMapViewController: NavigationMapViewDelegate {
     }
 
-    func navigationMapView(_ mapView: NavigationMapView, routeStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer? {
-        return delegate?.navigationMapView(mapView, routeStyleLayerWithIdentifier: identifier, source: source)
```
Why does this PR remove the ability to customize the route line?

bsudekum replied:
Fixed.

1ec5 commented on MapboxCoreNavigation/RouteController.swift:
```diff
@@ -22,8 +22,8 @@ public protocol RouteControllerDelegate: class {
-    @objc(routeController:shouldIncrementLegWhenArrivingAtWaypoints:)
-    optional func routeController(_ routeController: RouteController, shouldIncrementLegWhenArrivingAtWaypoints waypoint: Waypoint) -> Bool
+    @objc(routeController:shouldIncrementLegWhenArrivingAtWaypoint:)
+    optional func routeController(_ routeController: RouteController, shouldIncrementLegWhenArrivingAtWaypoint waypoint: Waypoint) -> Bool
```
The Swift name should be `routeController(_:shouldIncrementLegWhenArrivingAt:)`. “Waypoint” is inferred from the argument’s type.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -142,20 +149,25 @@ open class NavigationMapView: MGLMapView {
+            let circles = navigationMapDelegate?.navigationMapView?(self, waypointStyleLayerWithIdentifier: waypointCircleIdentifier, source: sourceShape) ?? routeWaypointCircleStyleLayer(identifier: waypointCircleIdentifier, source: sourceShape)
             
-            let symbols = navigationMapDelegate?.navigationMapView?(self, routeWaypointSymbolStyleLayerWithIdentifier: waypointSymbolIdentifier, source: sourceShape) ?? routeWaypointSymbolStyleLayer(identifier: waypointSymbolIdentifier, source: sourceShape)
+            let symbols = navigationMapDelegate?.navigationMapView?(self, waypointSymbolStyleLayerWithIdentifier: waypointSymbolIdentifier, source: sourceShape) ?? routeWaypointSymbolStyleLayer(identifier: waypointSymbolIdentifier, source: sourceShape)
```
If the developer wants the waypoint to be marked by a circle with no letter, they’d return `nil` for `navigationMapView(_:waypointSymbolStyleLayerWithIdentifier:source:)`, but then we’d go ahead and add a letter anyways.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -197,7 +209,7 @@ open class NavigationMapView: MGLMapView {
     }
     
-    func shape(for waypoints: [Waypoint]) -> MGLShape? {
+    func shapes(for waypoints: [Waypoint]) -> MGLShape? {
```
The name of this method implies that it returns an array of set of MGLShape objects. However, it returns only a single MGLShape object.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -276,6 +288,129 @@ open class NavigationMapView: MGLMapView {
         return lineCasing
     }
+    
+    func hideArrow(_ routeProgress: RouteProgress) {
```
This method doesn’t hide the arrow.

bsudekum replied:
Wow

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -298,5 +433,5 @@ public protocol NavigationMapViewDelegate: class  {
     
     @objc(navigationMapView:shapeDescribingWaypoints:)
-    optional func navigationMapView(_ mapView: NavigationMapView, shapeFor waypoints: [Waypoint]) -> MGLShape?
+    optional func navigationMapView(_ mapView: NavigationMapView, shapesFor waypoints: [Waypoint]) -> MGLShape?
```
Similarly, this method name now implies that the method returns an array or set of MGLShape objects, which isn’t the case.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -75,12 +75,19 @@ public protocol NavigationViewControllerDelegate {
+     If this method is unimplemented, the navigation map view draws the route line using an `MGLLineStyleLayer`.
+     */
+    @objc optional func navigationMapView(_ mapView: NavigationMapView, routeWaypointSymbolStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
     
```
Then remove runtime styling in favor of annotations, because that’s the usual way people add markers to the map. The main reason to expose this customization hook is to allow the developer to make the destination marker match the annotation on the main map in their application (the map that isn’t NavigationMapView).

1ec5 replied:
Then remove runtime styling in favor of annotations, because that’s the usual way people add markers to the map. The main reason to expose this customization hook is to allow the developer to make the destination marker match the annotation on the main map in their application (the map that isn’t NavigationMapView).

1ec5 replied:
Exposing `navigationMapView(_:viewForAnnotation:waypoint:)` would also fix #213.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -105,7 +98,12 @@ public protocol NavigationViewControllerDelegate {
     
-    @objc optional func navigationMapView(_ mapView: NavigationMapView, shapeFor waypoints: [Waypoint]) -> MGLShape?
+    
+    @objc optional func navigationMapView(_ mapView: NavigationMapView, waypointStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer?
```
Remember to document these new public methods.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -142,20 +149,25 @@ open class NavigationMapView: MGLMapView {
         }
         
-        let source = navigationMapDelegate?.navigationMapView?(self, shapeFor: routeProgress.remainingWaypoints) ?? shape(for: Array(routeProgress.remainingWaypoints.dropLast()))
+        let source = navigationMapDelegate?.navigationMapView?(self, shapesFor: routeProgress.remainingWaypoints) ?? shapes(for: Array(routeProgress.remainingWaypoints.dropLast()))
```
This isn’t what I meant by https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r132097275. The line below says `let s`. A single-letter variable should be a red flag. It was previously called `source`, but you changed it to `s` in order for this variable to be called `source`. Instead, call them each `<something>Source`.

1ec5 commented on MapboxNavigation/NavigationViewController.swift:
```diff
@@ -382,10 +392,6 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
     }
     
-    func navigationMapView(_ mapView: NavigationMapView, routeStyleLayerWithIdentifier identifier: String, source: MGLSource) -> MGLStyleLayer? {
-        return navigationDelegate?.navigationMapView?(mapView, routeStyleLayerWithIdentifier: identifier, source: source)
```
Per https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r132099394, this PR still removes the ability to customize the route line.

1ec5 commented on MapboxNavigation/RouteTableViewController.swift:
```diff
@@ -66,20 +66,41 @@ class RouteTableViewController: UIViewController {
+        let legToFrom = leg.name.components(separatedBy: ", ")
+        
+        if let destinationName = destinationName?.nonEmptyString, legToFrom.count > 1 {
+            return String.localizedStringWithFormat(NSLocalizedString("WAYPOINT_DESTINATION_VIA_WAYPOINTS_FORMAT", bundle: .mapboxNavigation, value: "%@, via %@", comment: "Format for displaying destination and intermediate waypoints; 1 = source ; 2 = destinations"), destinationName, leg.name.replacingOccurrences(of: ", ", with: " and "))
```
The word “and” should be localizable.

bsudekum commented:
Updated the route line and made the non active leg opaque:

![simulator screen shot aug 10 2017 2 16 05 pm](https://user-images.githubusercontent.com/1058624/29192533-9addd24a-7dd6-11e7-8e8e-24a95c5eeec4.png)

The stacked opacity at geometry intersections is due to https://github.com/mapbox/mapbox-gl-native/issues/1771 😢 

When the user enters the new leg, the route is updated accordingly.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -125,49 +137,141 @@ open class NavigationMapView: MGLMapView {
+        for (waypointIndex, waypoint) in waypoints.enumerated() {
+            let feature = MGLPointFeature()
+            feature.coordinate = waypoint.coordinate
+            feature.attributes = [ "name": letters[waypointIndex] ]
```
If we’re going to label each waypoint with a letter, then the set of letters should be localizable, because each language has its own order. (For example, in Vietnamese it begins “A Ă Â B C D Đ E Ê …”. In Simplified Chinese, it begins with “万千百十…”.)

It would be great if we could use `Locale.exemplarCharacterSet`, but unfortunately there isn’t a way to enumerate a CharacterSet’s members while preserving order. We’ll need a single localizable string that contains a space- or comma-delimited list of exemplar characters.

bsudekum commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -171,45 +172,72 @@ open class NavigationMapView: MGLMapView {
+                let polyline = MGLPolylineFeature(coordinates: congestionSegment.0, count: UInt(congestionSegment.0.count))
+                polyline.attributes["congestion"] = String(describing: congestionSegment.1)
+                if let legIndex = legIndex {
+                    polyline.attributes[isOpaqueIdentifier] = index != legIndex
```
🤔  I'm wondering if it would get confusing to the user if we added a transparency to both non active legs and completed routes.

bsudekum commented:
@1ec5 this is ready for another round of reviews.

bsudekum commented:
More reviews are welcome if anyone is up for it @1ec5 @frederoni.

bsudekum commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -154,49 +166,168 @@ open class NavigationMapView: MGLMapView {
+        for (waypointIndex, waypoint) in waypoints.enumerated() {
+            let feature = MGLPointFeature()
+            feature.coordinate = waypoint.coordinate
+            feature.attributes = [ "name": letters[waypointIndex] ]
```
This seems a bit unreadable IMO:

```swift
for (waypointIndex, waypoint) in waypoints.enumerated() {
  let feature = MGLPointFeature()
  feature.coordinate = waypoint.coordinate
  feature.attributes = [ "name": String.localizedStringWithFormat(NSLocalizedString("ALPHABET", bundle: .mapboxNavigation, value: "ABCDEFGHIJKLMNOPQRSTUVWXYZ", comment: "Format string for alphabet;")).characters.map { String(describing: $0) }[waypointIndex] ]
  features.append(feature)
}
```

bsudekum replied:
What if one waypoint has a name and the other does not? So it'd look like `B` (for picking up bobby) and then a waypoint with no name `B` again? Some sort of order makes more sense to me here.

1ec5 replied:
Ah, the code I suggested in https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r132049320 was much more readable, but it no longer makes sense with the localization that we’re doing. How about this:

```swift
let letters = String.localizedStringWithFormat(NSLocalizedString("ALPHABET", bundle: .mapboxNavigation, value: "ABCDEFGHIJKLMNOPQRSTUVWXYZ", comment: "Base letters in the alphabet for labeling waypoints, separated by spaces")).components(separatedBy: " ")

// then in the for loop
feature.attributes = ["name": letters[waypointIndex]]
```

1ec5 replied:
If any of the waypoints has a name, I think it would be better to label the named ones with initials and leave the unnamed ones… unlabeled.

1ec5 commented on MapboxNavigation/NavigationMapView.swift:
```diff
@@ -13,6 +14,17 @@ open class NavigationMapView: MGLMapView {
+    let arrowLayerStrokeIdentifier = "arrowStrokeLayer"
+    let arrowCasingSymbolLayerIdentifier = "arrowCasingSymbolLayer"
+    let arrowSymbolSourceIdentifier = "arrowSymbolSource"
+    let isCurrentLeg = "isCurrentLeg"
```
https://github.com/mapbox/mapbox-navigation-ios/pull/270#discussion_r134897473 got collapsed for some reason: the name `isCurrentLeg` implies that it’s a Boolean. Since this string is being used as an attribute name, call it `currentLegAttribute` but give it the value `isCurrentLeg`.

hervouinc commented:
🍾🥂

johnwallco commented:
When will multiple waypoints be available?
